plugins/machina-tools.js

```diff
@@ -0,0 +1,554 @@
+    // a big cache, and we don't have complex checking on adds.
+    delete machinaTools.portalCircles[guid];
+  }
+};
+
+machinaTools.removeConflictZone = function (guid) {
+  delete machinaTools.conflictZones[guid];
+};
+/**
+ * Reacts to a portal being added or removed.
+ */
+machinaTools.portalAdded = function (data) {
+  // Draw the circle if the team of the portal is Machina.
+  data.portal.on('add', function () {
+    // debugger;
+    // if (TEAM_NAMES[this.options.team] != undefined) {
+    if (window.TEAM_NAMES[this.options.team] === window.TEAM_NAME_MAC) {
+      machinaTools.drawPortalExclusion(this.options.guid);
+    }
+    // }
+  });
+
+  // Remove all circles if they exist, since the team may have changed.
+  data.portal.on('remove', function () {
+    machinaTools.removePortalExclusion(this.options.guid);
+    // machinaTools.removeConflictZone(this.options.guid);
+  });
+};
+
+/**
+ * Hides or shows the circle display layer as requested.
+ */
+machinaTools.showOrHideMachinaLevelUpRadius = function () {
+  if (machinaTools.zoomLevelHasPortals()) {
+    // Add the circle layer back to the display layer if necessary, and remove the disabled mark.
+    if (!machinaTools.displayLayer.hasLayer(machinaTools.circleDisplayLayer)) {
+      machinaTools.displayLayer.addLayer(machinaTools.circleDisplayLayer);
+      $('.leaflet-control-layers-list span:contains("Machina Level Up Link Radius")').parent('label').removeClass('disabled').attr('title', '');
+    }
+  } else {
+    // Remove the circle layer from the display layer if necessary, and add the disabled mark.
+    if (machinaTools.displayLayer.hasLayer(machinaTools.circleDisplayLayer)) {
+      machinaTools.displayLayer.removeLayer(machinaTools.circleDisplayLayer);
+      $('.leaflet-control-layers-list span:contains("Machina Level Up Link Radius")')
+        .parent('label')
+        .addClass('disabled')
+        .attr('title', 'Zoom in to show those.');
+    }
+  }
+};
+
+machinaTools.showOrHideMachinaConflictArea = function () {
+  if (!machinaTools.conflictLayer.hasLayer(machinaTools.conflictAreaLayer)) {
+    machinaTools.conflictLayer.addLayer(machinaTools.conflictAreaLayer);
+    console.log('conflictAreaLayer activated');
+    $('.leaflet-control-layers-list span:contains("Machina Conflict Area)').parent('label').removeClass('disabled').attr('title', '');
+  } else {
+    machinaTools.conflictLayer.removeLayer(machinaTools.conflictAreaLayer);
+    console.log('conflictAreaLayer disabled');
+    $('.leaflet-control-layers-list span:contains("Machina Conflict Area")').parent('label').addClass('disabled').attr('title', 'Zoom in to show those.');
+  }
+};
+
+machinaTools.guessLevelByRange = function (linkLength) {
+  for (var level = 0; level <= 8; ++level) {
+    if (window.LINK_RANGE_MAC[level] >= linkLength) {
+      return level;
+    }
+  }
+  return 0;
+};
+
+machinaTools.linkAdded = function (data) {
+  data.link.on('add', function () {
```

Comment:
this occurs when disabling/re-enabling links/machina layer
you can directly use the `data.link` as machina links don't change